controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,373 @@
+
+	NudgeProcessedAnnotationName = "build.appstudio.openshift.io/component-nudge-processed"
+	NudgeFinalizer               = "build.appstudio.openshift.io/build-nudge-finalizer"
+	FailureCountAnnotationName   = "build.appstudio.openshift.io/build-nudge-failures"
+
+	// AlreadyNudgedAnnotationName this is populated in the case of an error when some components were successful
+	AlreadyNudgedAnnotationName = "build.appstudio.openshift.io/nudged-components"
+
+	ComponentNudgedEventType      = "ComponentNudged"
+	ComponentNudgeFailedEventType = "ComponentNudgeFailed"
+	MaxAttempts                   = 3
+	FailureRetryTime              = time.Minute * 5 // We retry after 5 minutes on failure
+)
+
+// ComponentDependencyUpdateReconciler reconciles a PipelineRun object
+type ComponentDependencyUpdateReconciler struct {
+	client.Client
+	Scheme         *runtime.Scheme
+	EventRecorder  record.EventRecorder
+	UpdateFunction UpdateComponentDependenciesFunction
+}
+
+// SetupController creates a new Integration reconciler and adds it to the Manager.
+func (r *ComponentDependencyUpdateReconciler) SetupWithManager(manager ctrl.Manager) error {
+	return setupControllerWithManager(manager, r)
+}
+
+// setupControllerWithManager sets up the controller with the Manager which monitors new PipelineRuns and filters
+// out pipelines we don't need
+func setupControllerWithManager(manager ctrl.Manager, reconciler *ComponentDependencyUpdateReconciler) error {
+	return ctrl.NewControllerManagedBy(manager).
+		For(&tektonapi.PipelineRun{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return true
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+
+				new, ok := e.ObjectNew.(*tektonapi.PipelineRun)
+				if !ok {
+					return false
+				}
+
+				// Only handle push events
+				if new.ObjectMeta.Annotations != nil && new.ObjectMeta.Annotations[PacEventTypeLabelName] != PacEventPushType {
+					return false
+				}
+
+				// Ensure the PipelineRun belongs to a Component
+				if new.ObjectMeta.Labels == nil || new.ObjectMeta.Labels[ComponentNameLabelName] == "" {
+					// PipelineRun does not belong to a Component
+					return false
+				}
+				// Ensure we have not finished processing
+				if new.ObjectMeta.Labels != nil && new.ObjectMeta.Labels[NudgeProcessedAnnotationName] != "" {
+					return false
+				}
+
+				if new.Status.CompletionTime == nil || new.Status.CompletionTime.IsZero() {
+					// Pipeline run is still in progress, only return true if we need to add the finalizer
+					return !controllerutil.ContainsFinalizer(new, NudgeFinalizer)
+				}
+				return true
+
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return controllerutil.ContainsFinalizer(e.Object, NudgeFinalizer)
+			},
+			GenericFunc: func(e event.GenericEvent) bool {
+				return false
+			},
+		})).
+		WithEventFilter(predicate.NewPredicateFuncs(IsBuildPushPipelineRun)).
```

Comment:
How about removing this call and applying `IsBuildPushPipelineRun` in the newly added predicates?

Reply:
ah, we don't need that now that we have the predicate.